### PR TITLE
refactor: simplify inner dependencies and structure for easier setup with and without fx

### PR DESCRIPTION
**Pull request by denopink**
- add ability for clients to setup an ancla client with only `anclafx.Provide()`
- simplify calls to setup ancla without fx
- simplify inner dependencies and structure

### CODE REVIEW

File: chrysom/basicClient.go [hunk 1]
```diff
@@ -52,18 +55,18 @@ type BasicClientConfig struct {
...
+
+	// PullInterval is how often listeners should get updates.
+	// (Optional). Defaults to 5 seconds.
+	PullInterval time.Duration
 }
 
 // BasicClient is the client used to make requests to Argus.
 type BasicClient struct {
-	client       *http.Client
+	client       HTTPClient
```
Comment (denopink) [inline chrysom/basicClient.go:69]: This allows custom http clients since servers like tr1d1um need it. ATM, the current setup forces the user to do some not so great things to accomplish the same thing.

File: webhookValidationConfig.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ type OptionsConfig struct {
 }
 
 // BuildURLChecker translates the configuration into url Checker to be run on the webhook.
-func (config *ValidatorConfig) BuildURLChecker() (*urlegit.Checker, error) {
```
Comment (denopink) [inline webhookValidationConfig.go:54]: I may move this to another pr, but this doesn't have to be a member (pointer) of ValidatorConfig. This will simplify the mental model.

File: chrysom/basicClient.go [hunk 3]
```diff
@@ -85,22 +88,46 @@ const (
...
+type HTTPClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type BasicClientIn struct {
+	fx.In
+
+	BasicClientConfig BasicClientConfig
+	// (Optional) Defaults to http.DefaultClient.
+	HTTPClient HTTPClient
```
Comment (denopink) [inline chrysom/basicClient.go:100]: This allows custom http clients since servers like tr1d1um need it. ATM, the current setup forces the user to do some not so great things to accomplish the same thing.

File: chrysom/metrics.go [hunk 4]
```diff
@@ -3,12 +3,6 @@
...
 package chrysom
 
-import (
-	"github.com/prometheus/client_golang/prometheus"
-	"github.com/xmidt-org/touchstone"
-	"go.uber.org/fx"
-)
-
 // Names
 const (
```
Comment (maurafortino) [inline chrysom/metrics.go:7]: is it necessary to have this file just for three constants? would it make sense to move them to another file?

File: service.go [hunk 5]
```diff
@@ -125,65 +95,65 @@ func (s *ClientService) GetAll(ctx context.Context) ([]Register, error) {
...
+}
+
+// ProvideService builds the Argus client service from the given configuration.
+func ProvideService(in ClientServiceIn) *ClientService {
+	return NewService(in.BasicClient)
+}
+
+// TODO: Refactor and move Watch and Listener related code to chrysom.
+
+type DefaultListenerOut struct {
```
Comment (maurafortino) [inline service.go:118]: missing fx.Out ?

File: service.go [hunk 6]
```diff
@@ -125,65 +95,65 @@ func (s *ClientService) GetAll(ctx context.Context) ([]Register, error) {
...
+	Watchers []Watch `group:"watchers,flatten"`
+}
+
+type DefaultListenersIn struct {
```
Comment (maurafortino) [inline service.go:122]: missing fx.In ?

File: chrysom/basicClient.go [hunk 7]
```diff
@@ -52,18 +55,18 @@ type BasicClientConfig struct {
 	// Bucket partition to be used by this client.
 	Bucket string
 
-	// HTTPClient refers to the client that will be used to send requests.
-	// (Optional) Defaults to http.DefaultClient.
-	HTTPClient *http.Client
-
 	// Auth provides the mechanism to add auth headers to outgoing requests.
 	// (Optional) If not provided, no auth headers are added.
 	Auth Acquirer
```
Comment (schmidtw) [inline chrysom/basicClient.go:60]: The http.Client and Acquirer should be replaced by a the `arrangehttp.ClientConfig`.  I think this lets you delete most of chrysom.

File: chrysom/basicClient.go (hunk 1 above)
Comment (schmidtw) [inline chrysom/basicClient.go:69]: A *http.Client is nice here instead of the interface.  For testing use httptest package.

File: chrysom/basicClient.go [hunk 8]
```diff
@@ -85,22 +88,46 @@ const (
 // Items is a slice of model.Item(s) .
 type Items []model.Item
 
+type HTTPClient interface {
```
Comment (schmidtw) [inline chrysom/basicClient.go:91]: Remove in favor of httptest package.

File: chrysom/basicClient.go (hunk 3 above)
Comment (schmidtw) [reply chrysom/basicClient.go:100]: I don't follow why having the configuration for this feature in the BasicClientConfig is worse than here.  This to me makes configuring the right http client harder.

File: chrysom/basicClient_test.go [hunk 9]
```diff
@@ -147,6 +144,20 @@ func TestSendRequest(t *testing.T) {
 			MockError: nil,
 			MockAuth:  "basic xyz",
 		},
+		{
+			Description: "Happy path with default http client",
+			Method:      http.MethodPut,
+			URL:         "http://argus-hostname.io",
```
Comment (schmidtw) [inline chrysom/basicClient_test.go:150]: Stay with something based on `example.com` because the url here could be a legitimate one & `example.com` is supposed to be reserved for testing.

File: chrysom/listenerClient.go [hunk 10]
```diff
@@ -39,15 +40,16 @@ const (
...
+	fx.In
 
-	// PullInterval is how often listeners should get updates.
-	// (Optional). Defaults to 5 seconds.
-	PullInterval time.Duration
+	// Listener fetches a copy of all items within a bucket on
+	// an interval based on `BasicClientConfig.PullInterval`.
+	// (Optional). If not provided, listening won't be enabled for this client.
+	Listener          Listener
+	Config            BasicClientConfig
```
Comment (schmidtw) [inline chrysom/listenerClient.go:50]: Having documentation for the fields would be nice.

File: webhookValidationConfig.go (hunk 2 above)
Comment (schmidtw) [reply webhookValidationConfig.go:54]: This change seems to add no value, why is it here?

**Review by schmidtw — CHANGES_REQUESTED** [review]

File: chrysom/basicClient.go (hunk 7 above)
Comment (denopink) [reply chrysom/basicClient.go:60]: I was thinking about using `arrangehttp.ClientConfig`, but I saw more trouble than it's worth (potentially having ancla's arrange version clash with other servers versions).

I was thinking having the user provide a http.client was a the amplifier path forward.

File: chrysom/basicClient.go (hunk 1 above)
Comment (denopink) [reply chrysom/basicClient.go:69]: Ah, that's such a great point!

File: chrysom/basicClient.go (hunk 3 above)
Comment (denopink) [reply chrysom/basicClient.go:100]: `Harder` was too strong of a word.

Having the http client in BasicClientConfig will require the user to unmarshal their `BasicClientConfig` and then open it up and then add their own http client. I didn't like this because it requires us to modify `BasicClientConfig` struct during dependency injection.

I thought it would be cleaner to have the user inject their own http.client with fx and have ancla use that.

File: chrysom/metrics.go (hunk 4 above)
Comment (denopink) [reply chrysom/metrics.go:7]: I think so. I may even move the touchstone options in here

File: service.go (hunk 5 above)
Comment (denopink) [reply service.go:118]: whops! ur right 🙂

File: service.go (hunk 6 above)
Comment (denopink) [reply service.go:122]: yup!

File: webhookValidationConfig.go (hunk 2 above)
Comment (denopink) [reply webhookValidationConfig.go:54]: It's only here because I accidentally commit it with the other changes. Fair, I'll trash them 🗑️

File: chrysom/basicClient_test.go (hunk 9 above)
Comment (denopink) [reply chrysom/basicClient_test.go:150]: I didn't know that about `example.com`, that's neat!

File: anclafx/provide.go [hunk 11]
```diff
@@ -23,17 +21,6 @@ func Provide() fx.Option {
...
-				Help: chrysom.WebhookListSizeGaugeHelp,
-			}),
-		touchstone.CounterVec(
-			prometheus.CounterOpts{
-				Name: chrysom.PollsTotalCounterName,
-				Help: chrysom.PollsTotalCounterHelp,
-			},
-			chrysom.OutcomeLabel,
-		),
+		chrysom.ProvideMetrics(),
```
Comment (denopink) [inline anclafx/provide.go:24]: Move the tochstone options to `chrysom.ProvideMetrics()`, better separation of concerns.

File: anclafx/provide.go [hunk 12]
```diff
@@ -12,8 +13,14 @@ const Module = "ancla"
 func Provide() fx.Option {
```
Comment (denopink) [inline anclafx/provide.go:13]: I'll add a fx test in `anclafx/provide_test.go`

**Review by johnabass — APPROVED** [review]
Looks like you addressed the one comment I had.  Good job!
